flink/src/main/java/org/apache/iceberg/flink/source/StreamingReaderOperator.java

```diff
@@ -0,0 +1,233 @@
+
+  private transient ListState<FlinkInputSplit> inputSplitsState;
+  private transient Queue<FlinkInputSplit> splits;
+
+  // This state is used to control that only one split is occupying the executor for record reading. If there're more
+  // splits coming, we will buffer them in flink's state. Once the executor is idle (the currentSplitState will be
+  // marked as IDLE), it will schedule one new split from buffered splits in flink's state to executor (the
+  // currentSplitState will be marked as RUNNING). After finished all records processing, the currentSplitState will
+  // be marked as IDLE again.
+  // NOTICE: all the reader and writer of this variable are the same thread, so we don't need extra synchronization.
+  private transient SplitState currentSplitState;
+
+  private StreamingReaderOperator(FlinkInputFormat format, ProcessingTimeService timeService,
+                                  MailboxExecutor mailboxExecutor) {
+    this.format = Preconditions.checkNotNull(format, "The InputFormat should not be null.");
+    this.processingTimeService = timeService;
+    this.executor = Preconditions.checkNotNull(mailboxExecutor, "The mailboxExecutor should not be null.");
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    super.initializeState(context);
+
+    // TODO Replace Java serialization with Avro approach to keep state compatibility.
+    // See issue: https://github.com/apache/iceberg/issues/1698
+    inputSplitsState = context.getOperatorStateStore().getListState(
+        new ListStateDescriptor<>("splits", new JavaSerializer<>()));
+
+    // Initialize the current split state to IDLE.
+    currentSplitState = SplitState.IDLE;
+
+    // Recover splits state from flink state backend if possible.
+    splits = Lists.newLinkedList();
+    if (context.isRestored()) {
+      int subtaskIdx = getRuntimeContext().getIndexOfThisSubtask();
+      LOG.info("Restoring state for the {} (taskIdx: {}).", getClass().getSimpleName(), subtaskIdx);
+
+      for (FlinkInputSplit split : inputSplitsState.get()) {
+        splits.add(split);
+      }
+    }
+
+    this.sourceContext = StreamSourceContexts.getSourceContext(
+        getOperatorConfig().getTimeCharacteristic(),
+        getProcessingTimeService(),
+        new Object(), // no actual locking needed
+        getContainingTask().getStreamStatusMaintainer(),
+        output,
+        getRuntimeContext().getExecutionConfig().getAutoWatermarkInterval(),
+        -1);
+
+    if (!splits.isEmpty()) {
+      enqueueProcessSplits();
+    }
+  }
+
+  @Override
+  public void processElement(StreamRecord<FlinkInputSplit> element) {
+    splits.add(element.getValue());
+    enqueueProcessSplits();
+  }
+
+  private void enqueueProcessSplits() {
+    if (currentSplitState == SplitState.IDLE) {
+      executor.execute(this::processSplits, this.getClass().getSimpleName());
+      currentSplitState = SplitState.RUNNING;
```

Comment:
Split state should be set before starting a concurrent operation right?

Reply:
Moving this line before line-130 is OK for me.

Reply:
If there aren't multiple threads, then I agree it doesn't matter much.